### DISCUSSION TITLE
when upload new object to bucket, can I set up priority to write to ssd disk if I have both ssd and hdd disks ?

### DISCUSSION THREAD

**Discussion by astrung** [seaweedfs/seaweedfs#4211] [Q&A] [answered]
I have 4 disks: 2 ssd and 2 hdd disks. I need to do following procedures:

1. Insert 4 disks into seaweedfs cluster as volumes. Tag each volume with its corresponding disk. So I will have 4 volumes: 2 volume servers are tagged as ssd disks, and 2 volume servers as 2 hdd disks
2. Create a new bucket: `my_bucket` through S3 API (I also have a filer and S3 server in seaweedfs cluster too)
3. **Write new object `X` to `my_bucket`. And I need to write it to ssd disk**
4. After 1 days, if object `X` doesn't get any changes, move it to hdd disks.

* By this following docker-compose file, I can achieve step 1, where I assign each volume with a tag(hdd or ssd)
[cluster-docker-compose.txt](https://github.com/seaweedfs/seaweedfs/files/10754549/cluster-docker-compose.txt)
* By using S3 API, I can create bucket.
* **But I don't know how to setup in step 3. When I put a new object, it is uploaded to a random volume, so it can be ssd or hdd disk. I need to put new object only in ssd disk. Please help me there - How can I do it ????**
* By using this command, I can move all of objects to ssd disk: `volume.tier.move -fromDiskType=hdd -toDiskType=ssd -quietFor=24h -fullPercent=0.001 -collectionPattern=newbucket3 -force`

Can we setup a priority to write to ssd disk in step 3 ? please help

**Comment by astrung** [comment]
sorry but can anyone help ???

**Comment by chrislusf** [comment]
use https://github.com/seaweedfs/seaweedfs/wiki/Path-Specific-Configuration#options to configure the disk type for a path.

**Reply by astrung** [reply]
@chrislusf thank you for help. I have achieved step 3 by setup this command in `weed shell`: `fs.configure -locationPrefix=/buckets/ssd_ -disk=ssd_bucket -apply`. Now all of object will be stored in ssd disk. 
And even after moving all volume to hdd disk by this command: `volume.tier.move -fromDiskType=ssd -toDiskType=hdd -quietFor=24h -fullPercent=0.001 -collectionPattern=ssd_bucket -force`, it still is in same bucket.

But now I have a new problem: **after moving volume to hdd disk, I see that my volumes don't have any replication**. In other word, I have 2 volumes in my ssd disk (1 base volume + 1 replication volume), by **moving with command `volume.tier.move`, replication volume is removed and my hdd disk only has 1 base volume**. Can command `volume.tier.move` have any options to keep my replication, or moving replication volume to hdd disk too ?

**Reply by chrislusf** [reply]
you can run a command to fix replication

**Reply by astrung** [reply]
@chrislusf sorry but what command ? I tried to run command `volume.tier.move -fromDiskType=hdd -toDiskType=ssd -quietFor=1s -fullPercent=0.001 -collection=benchmark -toReplication 100 -force`, but still ssd disks doesn't have any replication volume. How can I fix it ? Please help

**Comment by chrislusf** [comment]
`volume.fix.replication`

**Reply by astrung** [reply]
@chrislusf 
I ran this command, but still get error. `error: replicas volume 4 mismatch in topology`. 
I don't change any configuration for replication or volume.
This is my log:

    > volume.tier.move -fromDiskType=ssd -toDiskType=hdd -quietFor=1s -fullPercent=0.001 -collectionPattern=ssd_bucket -force
    collect ssd volumes quiet for: 1 seconds
    tier move volumes: [4]
    moving volume 4 from volume2:8082 to volume1:8080 with disk type hdd ...
    markVolumeReadonly 4 on volume4:8084 ...
    markVolumeReadonly 4 on volume2:8082 ...
    2023/02/21 07:57:56 copying volume 4 from volume2:8082 to volume1:8080
    2023/02/21 07:57:56 tailing volume 4 from volume2:8082 to volume1:8080
    2023/02/21 07:58:08 deleting volume 4 from volume2:8082
    2023/02/21 07:58:08 moved volume 4 from volume2:8082 to volume1:8080
    > volume.fix.replication
    replicating volume 4 100 from volume1:8080 to dataNode volume3:8083 ...
    the number of locations for volume 4 has not increased yet, let's wait
    error: replicas volume 4 mismatch in topology

And this is log for 4 volume (2ssd, 2 hdd):

    Topology volumeSizeLimit:100 MB ssd(volume:2/4576 active:2 free:4574 remote:0) hdd(volume:6/4576 active:6 free:4570 remote:0)
      DataCenter dc1 ssd(volume:0/0 active:0 free:0 remote:0) hdd(volume:3/2288 active:3 free:2285 remote:0)
        Rack v1 hdd(volume:3/2288 active:3 free:2285 remote:0)
          DataNode volume1:8080 hdd(volume:3/2288 active:3 free:2285 remote:0)
            Disk hdd(volume:3/2288 active:3 free:2285 remote:0)
              volume id:1  size:448  file_count:1  replica_placement:100  version:3  modified_at_second:1676966265 
              volume id:2  size:976  file_count:1  replica_placement:100  version:3  modified_at_second:1676966205 
              volume id:4  size:69016  collection:"ssd_bucket"  file_count:1  replica_placement:100  version:3  modified_at_second:1676966207 
            Disk hdd total size:70440 file_count:3 
          DataNode volume1:8080 total size:70440 file_count:3 
        Rack v1 total size:70440 file_count:3 
      DataCenter dc1 total size:70440 file_count:3 
      DataCenter dc2 ssd(volume:1/2288 active:1 free:2287 remote:0) hdd(volume:0/0 active:0 free:0 remote:0)
        Rack v2 ssd(volume:1/2288 active:1 free:2287 remote:0)
          DataNode volume2:8082 ssd(volume:1/2288 active:1 free:2287 remote:0)
            Disk ssd(volume:1/2288 active:1 free:2287 remote:0)
              volume id:3  size:8  collection:"ssd_bucket"  replica_placement:100  version:3  modified_at_second:1676966207  disk_type:"ssd" 
            Disk ssd total size:8 file_count:0 
          DataNode volume2:8082 total size:8 file_count:0 
        Rack v2 total size:8 file_count:0 
      DataCenter dc2 total size:8 file_count:0 
      DataCenter dc3 hdd(volume:3/2288 active:3 free:2285 remote:0) ssd(volume:0/0 active:0 free:0 remote:0)
        Rack v3 hdd(volume:3/2288 active:3 free:2285 remote:0)
          DataNode volume3:8083 hdd(volume:3/2288 active:3 free:2285 remote:0)
            Disk hdd(volume:3/2288 active:3 free:2285 remote:0)
              volume id:1  size:448  file_count:1  replica_placement:100  version:3  modified_at_second:1676966265 
              volume id:2  size:976  file_count:1  replica_placement:100  version:3  modified_at_second:1676966205 
              volume id:4  size:69016  collection:"ssd_bucket"  file_count:1  replica_placement:100  version:3  modified_at_second:1676966207 
            Disk hdd total size:70440 file_count:3 
          DataNode volume3:8083 total size:70440 file_count:3 
        Rack v3 total size:70440 file_count:3 
      DataCenter dc3 total size:70440 file_count:3 
      DataCenter dc4 ssd(volume:1/2288 active:1 free:2287 remote:0) hdd(volume:0/0 active:0 free:0 remote:0)
        Rack v4 ssd(volume:1/2288 active:1 free:2287 remote:0)
          DataNode volume4:8084 ssd(volume:1/2288 active:1 free:2287 remote:0)
            Disk ssd(volume:1/2288 active:1 free:2287 remote:0)
              volume id:3  size:8  collection:"ssd_bucket"  replica_placement:100  version:3  modified_at_second:1676966207  disk_type:"ssd" 
            Disk ssd total size:8 file_count:0 
          DataNode volume4:8084 total size:8 file_count:0 
        Rack v4 total size:8 file_count:0 
      DataCenter dc4 total size:8 file_count:0 
    total size:140896 file_count:6

**Comment by chrislusf** [comment, marked as answer]
try this:
```
volume.fix.replication -retry=5
```

**Reply by astrung** [reply]
thank you. it work
